mypy/semanal.py

```diff
@@ -1954,6 +1954,15 @@ class Foo(Bar, Generic[T]): ...
             del base_type_exprs[i]
         tvar_defs: list[TypeVarLikeType] = []
         for name, tvar_expr in declared_tvars:
+            tvar_expr_default = tvar_expr.default
+            if isinstance(tvar_expr_default, UnboundType):
+                # Assumption here is that the names cannot be duplicated
```

Comment:
Unfortunately that's not a safe assumption

Reply:
Do you have an example I could try? Mypy will already complain about this here and reject the second `TypeVar` call.
```py
T1 = TypeVar("T1", default=int)
T1 = TypeVar("T1", default=str)  # error
```

```
error: Cannot redefine "T1" as a type variable
error: Invalid assignment target
error: "int" not callable
```

Reply:
That results in the same error message. The second `TypeVar` call would just be ignored. The logic here would still as it only compares the actual TypeVar name `type_var_name = fullname.rpartition(".")[2]`. Not the complete `fullname`.

E.g. the comparison is with `T1` not `__main__.T1`.

Reply:
I thought something like this might fail:

```
[case testTypeVarDefaultsMultipleFiles]

from typing import Generic, TypeVar
from file2 import T as T2

T = TypeVar('T')

class Gen(Generic[T, T2]):
    pass

def func(a: Gen, b: Gen[str], c: Gen[str, str]) -> None:
    reveal_type(a)  # N: Revealed type is "__main__.Gen[Any, builtins.int]"
    reveal_type(b)  # N: Revealed type is "__main__.Gen[builtins.str, builtins.int]"
    reveal_type(c)  # N: Revealed type is "__main__.Gen[builtins.str, builtins.str]"

[file file2.py]
from typing import TypeVar

T = TypeVar('T', default=int)
```

But this passes as expected, and I couldn't find a variation that fails. So I suppose this is fine.

Reply:
This variation fails:

```
[case testTypeVarDefaultsMultipleFiles]

from typing import Generic, TypeVar
from file2 import T as T2

T = TypeVar('T', default=T2)

class Gen(Generic[T2, T]):
    pass

def func(a: Gen, b: Gen[str], c: Gen[str, float]) -> None:
    reveal_type(a)  # N: Revealed type is "__main__.Gen[builtins.int, builtins.int]"
    reveal_type(b)  # N: Revealed type is "__main__.Gen[builtins.str, builtins.str]"
    reveal_type(c)  # N: Revealed type is "__main__.Gen[builtins.str, builtins.float]"

[file file2.py]
from typing import TypeVar

T = TypeVar('T', default=int)
```

```
Expected:
  main:11: note: Revealed type is "__main__.Gen[builtins.int, builtins.int]" (diff)
  main:12: note: Revealed type is "__main__.Gen[builtins.str, builtins.str]" (diff)
  main:13: note: Revealed type is "__main__.Gen[builtins.str, builtins.float]"
Actual:
  main:11: note: Revealed type is "__main__.Gen[builtins.int, T2?]" (diff)
  main:12: note: Revealed type is "__main__.Gen[builtins.str, T2?]" (diff)
  main:13: note: Revealed type is "__main__.Gen[builtins.str, builtins.float]"
```

Not code I'd expect anyone to actually write, but it does seem like it should be legal.

Reply:
Thanks for the test, that helped! Tbh I could also have though about it earlier but it's actually quite simple to add a name lookup here 😅